tests/core/channels/test_botframework.py

```diff
@@ -0,0 +1,155 @@
+    jwk["kid"] = "key_id"
+
+    openid_config_mock_response = mock.Mock()
+    openid_config_mock_response.raise_for_status = mock.Mock()
+    openid_config_mock_response.json.return_value = MS_OPENID_CONFIG_RESPONSE
+
+    openid_keys_mock_response = mock.Mock()
+    openid_keys_mock_response.raise_for_status = mock.Mock()
+    openid_keys_mock_response.json.return_value = {"keys": [jwk]}
+
+    mock_requests.side_effect = [openid_config_mock_response, openid_keys_mock_response]
+
+    return BotFrameworkInput("app_id", "app_password")
+
+
+def test_successful_jwt_signature_verification(
+    bot_framework_input: BotFrameworkInput,
+    rsa_private_key: RSAPrivateKey,
+):
+    encoded = jwt.encode(
+        {
+            "serviceurl": "https://webchat.botframework.com/",
+            "nbf": int(time.time()),
+            "exp": int(time.time()) + DAY_IN_SECONDS,
+            "iss": "https://api.botframework.com",
+            "aud": "app_id",
+        },
+        rsa_private_key,
+        algorithm="RS256",
+        headers={"kid": "key_id", "alg": "RS256"},
+    )
+
+    resp = bot_framework_input._validate_auth(f"Bearer {encoded}")
+    assert resp is None
+
+
+@mock.patch("requests.get")
+def test_jwk_is_updated_daily(mock_requests: Mock):
+    with freezegun.freeze_time("2012-01-14 08:00:00"):
+        bot_framework_input = BotFrameworkInput("app_id", "app_password")
+        # Two calls at the beginning - on to retrieve open ID metadata,
+        # the second one to actually get the keys.
+        assert mock_requests.call_count == 2
+
+        bot_framework_input._validate_auth("Bearer token_invalid")
+        assert mock_requests.call_count == 2
+
+    with freezegun.freeze_time("2012-01-14 11:00:00"):
+        bot_framework_input._validate_auth("Bearer token_invalid")
+        assert mock_requests.call_count == 2
+
+    with freezegun.freeze_time("2012-01-15 11:00:00"):
+        bot_framework_input._validate_auth("Bearer token_invalid")
+        assert mock_requests.call_count == 4
+
+
+def test_validate_auth_returns_unauthorized_for_invalid_jwt_token(
+    bot_framework_input: BotFrameworkInput,
+    rsa_private_key: RSAPrivateKey,
+):
+    encoded = jwt.encode(
+        {
+            "serviceurl": "https://webchat.botframework.com/",
+            "nbf": int(time.time()) + DAY_IN_SECONDS,
+            "exp": int(time.time()) + 2 * DAY_IN_SECONDS,
+            "iss": "https://api.botframework.com",
+            "aud": "app_id",
+        },
+        rsa_private_key,
+        algorithm="RS256",
+        headers={"kid": "key_id", "alg": "RS256"},
+    )
+
+    resp = bot_framework_input._validate_auth(f"Bearer {encoded}")
+    assert resp is not None
+    assert resp.status == HTTPStatus.UNAUTHORIZED
+
+
+def test_validate_auth_returns_unauthorized_for_absent_header(
+    bot_framework_input: BotFrameworkInput,
+):
+    resp = bot_framework_input._validate_auth(None)
+    assert resp is not None
```

Comment:
I don't think asserting the content of these texts provides any value, since they would always be discarded by the Azure bot framework, which is what calls these in the end. But no strong opinion, can add those if you think it makes sense 🙂

Reply:
I should have said these are for unit test coverage, rather than the correctness of the code.